### PR TITLE
Add timeouts to eth client

### LINKED ISSUE #2207: Improve Feed PubSub mechanism
## Abstract

In `go-livepeer` we use the in-process PubSub mechanism, based on the [event.Feed structure](https://pkg.go.dev/github.com/ethereum/go-ethereum/event#Feed), which distribute events between different goroutines. We use it with an assumption that it's async, but in reality, it's a blocking mechanism. This spec describes a set of activities to fix this issue.

## Motivation

There were a number of bugs reported, which most probably are related to this issue:
- https://github.com/livepeer/go-livepeer/issues/2152
- https://github.com/livepeer/go-livepeer/issues/2079

A sample scenario to experience the issue is to start an orchestrator and send a transaction which is not been minted for a long time. Then one of the PubSub subscribers waits on [client.CheckTx](https://github.com/livepeer/go-livepeer/blob/c2c73fa6f24473311766baee83e2762c87ef708d/eth/client.go#L870), which blocks the PubSub sender, which in result makes the orchestrator unable to do anything, even transcoding a segment.

## Proposed Solution

This diagram presents how we use the Feed PubSub mechanism in which each feed is a PubSub topic.

image

If anything blocks in any subscriber, it can block the whole graph. Why is it so?

The [Feed subscription mechanism](https://pkg.go.dev/github.com/ethereum/go-ethereum/event#Feed.Subscribe) works with Go channels and when the channel is full, the sender is blocked. The recommendation from the doc is:

> The channel should have ample buffer space to avoid blocking other subscribers. Slow subscribers are not dropped.

However, we have these 2 issues:

1. Subscribers execute blocking operations
2. Blocking operations don't use any deadline/timeout

As an example, we can see [client.CheckTx() in rewardservice](https://github.com/livepeer/go-livepeer/blob/v0.5.26/eth/rewardservice.go#L94). It's blocking and does not use any timeout.

We should address both of these issues, so make sure that each subscriber does not block and add timeouts for all blocking operations.

## Implementation Tasks and Considerations

1. Execute subscribers' blocking operations in separate goroutines
   - PR: #2208
2. Add timeouts for all blocking operations

## Testing Tasks and Considerations

Testing scenarios:
1. Insert `time.Sleep(1 * time.Hour)` into `CheckTx()` and other operations and check that the transcoding is not blocked
2. Check that no goroutines are blocked forever by inserting `time.Sleep(1 * time.Hour)` and checking `runtime.NumGoroutine()`

## Known Unknowns

Consider checking the following aspects:
1. Check O transcoding stability while executing a transaction:
   - I noticed that B retries a segment when O is in the process of executing a transaction
2. Race condition between executing a transaction and calling `CheckTx()`
   - `CheckTx()` is always executed after the transaction and technically the receipt of a transaction can be published before `CheckTx()` is called
   - it's a very corner case and should not happen in practice
3. Check the scenario in which the transaction is never minted
   - Make sure that the next transactions are not blocked
   - If for some reason a transaction is not minted, then we should probably abort it

## Alternatives

1\. Make `feed.Send()` async

Instead of making the subscribers handle events in the async fashion, we could send events in an async manner.

```golang
go func() {
  feed.Send()
}
```

This would actually be the simplest solution. However, this may not work, since the feed graph will still block, only `send()` does not block. It's also acting against the idea of the `event.Feed` mechanism.

2\. Use a different PubSub solution

We could find another library for the in-process PubSub which is really async. However, I belive it's an overkill at this point.

## Additional Context

Related GH Issues:
- https://github.com/livepeer/go-livepeer/issues/2168
- https://github.com/livepeer/go-livepeer/issues/2152
- https://github.com/livepeer/go-livepeer/issues/2079

### PR DESCRIPTION
**Pull request by leszko**
**What does this pull request do? Explain your changes. (required)**

Add timeouts for `CheckTx()` function and all ETH RPC calls.

Note that the timeout values are quite conservative, by default:
- 5 min for `CheckTx()`
- 10 min for every ETH RPC call

I don't think we need smaller numbers, because in the correct flow of action, these timeouts should never actually be reached.

**Specific updates (required)**
- Add timeout for the `CheckTx()` execution
- Add timeout for all ETH RPC (contracts) calls
- Refactor client to not include contract sessions anymore

**How did you test each of these updates (required)**

- Checked that using eth client and sending transactions still work fine
- Change `CheckTxTimeout` to 1 ns and check that it always times out
- Change `ethRpcTimeout` to 1 ns and check that all ETH RPC requests fail

**Does this pull request close any open issues?**

fix #2207

**Checklist:**

- [x] Read the [contribution guide](./doc/contributing.md)
- [x] `make` runs successfully
- [x] All tests in `./test.sh` pass
- [ ] README and other documentation updated
- [x] [Pending changelog](./CHANGELOG_PENDING.md) updated

### CODE REVIEW

File: eth/client.go [hunk 1]
```diff
@@ -139,23 +140,23 @@ type client struct {
...
-	controllerSess          *contracts.ControllerSession
-	livepeerTokenSess       *contracts.LivepeerTokenSession
-	serviceRegistrySess     *contracts.ServiceRegistrySession
-	bondingManagerSess      *contracts.BondingManagerSession
-	ticketBrokerSess        *contracts.TicketBrokerSession
-	roundsManagerSess       *contracts.RoundsManagerSession
-	minterSess              *contracts.MinterSession
-	livepeerTokenFaucetSess *contracts.LivepeerTokenFaucetSession
+	// Contracts
+	controller          *contracts.Controller
```
Comment (victorges) [inline eth/client.go:144]: Like the refactor, but is this related to the timeouts change? Seems to be the bulk of the PR here.

File: eth/client.go [hunk 2]
```diff
@@ -1101,6 +1089,8 @@ func (c *client) CheckTx(tx *types.Transaction) error {
 
 	for {
 		select {
+		case <-time.After(c.checkTxTimeout):
```
Comment (victorges) [inline eth/client.go]: I believe you actually need to create the `time.After` timer only once, outside the `for` loop, otherwise you will reset the timer for every receipt that you get even if not related to the tx you're checking.

Comment (leszko) [reply eth/client.go]: Updated in a354f7148d50b275912ac379f405d2f7dd9124ca

File: eth/client.go (hunk 1 above)
Comment (leszko) [reply eth/client.go:144]: It is related because before adding the timeout, we actually needed to store sessions here (`CallOpts` was encapsulated in the session structs). Now, we don't need sessions, because we always provide our `CallOpts`.

Comment (victorges) [reply eth/client.go:144]: Cool! Thanks for clarification

**Review by victorges — APPROVED** [review]
LGTM!

File: cmd/devtool/devtool.go [hunk 3]
```diff
@@ -199,6 +199,7 @@ func ethSetup(ethAcctAddr, keystoreDir string, isBroadcaster bool) {
 		GasPriceMonitor:    gpm,
 		TransactionManager: tm,
 		Signer:             types.LatestSignerForChainID(chainID),
+		CheckTxTimeout:     ethTxTimeout,
```
Comment (yondonfu) [inline cmd/devtool/devtool.go]: WDYT about setting this in the same way as in `livepeer.go` for consistency? i.e. multiply the timeout by the max replacements?

File: eth/client.go [hunk 4]
```diff
@@ -39,6 +39,7 @@ var (
 	ErrReplacingMinedTx   = fmt.Errorf("trying to replace already mined tx")
 	ErrCurrentRoundLocked = fmt.Errorf("current round locked")
 	ErrMissingBackend     = fmt.Errorf("missing Ethereum client backend")
+	ethRpcTimeout         = 10 * time.Minute
```
Comment (yondonfu) [inline eth/client.go]: I know this was the original timeout value, but seems like this should be lower than 10 minutes for receiving a response from a RPC call...maybe 2 minutes?

File: eth/client_ticketbroker.go [hunk 5]
```diff
@@ -26,7 +26,7 @@ func (c *client) FundDeposit(amount *big.Int) (*types.Transaction, error) {
 	opts := c.transactOpts()
 	opts.Value = amount
```
Comment (yondonfu) [inline eth/client_ticketbroker.go:27]: This was already here before this PR, but since `transactOpts` is stored on `client` and `c.transactOpts()` returns a pointer to that field does the update to the `Value` field in the next line get reflected in that field such that the next call to `c.transactOpts()` would return a pointer to a `bind.TransactOpts` with the `Value` field set (which we wouldn't want)?

File: cmd/devtool/devtool.go (hunk 3 above)
Comment (leszko) [reply cmd/devtool/devtool.go]: Yes, updated in ee6a8c16ee301b9d3eb1d6decd4f5280a9dca799.

You also made me realize that in `livepeer.go` it should actually be `txTimeout * (maxTxReplacements + 1)`, not `txTimeout * maxTxReplacements`. Updated as well.

File: eth/client.go (hunk 4 above)
Comment (leszko) [reply eth/client.go]: Yeah, 2 min sounds good to me. Updated in cdfb8d15dc15300483eaed64a2e8f4d9bf7b796e

File: eth/client_ticketbroker.go (hunk 5 above)
Comment (leszko) [reply eth/client_ticketbroker.go:27]: Actually, that's not the case, because `c.transactOpts()` returns a copy of `TransactOpts`. The copying happens [here](https://github.com/leszko/go-livepeer/blob/cdfb8d15dc15300483eaed64a2e8f4d9bf7b796e/eth/client.go#L368).

**Review by leszko — COMMENTED** [review]
@yondonfu Addressed your comments. PTAL

File: eth/client_ticketbroker.go (hunk 5 above)
Comment (yondonfu) [reply eth/client_ticketbroker.go:27]: Aha I see.

The field `c.transactOpts` is a value so `opts := c.transactOpts` will copy the value and assign it to `opts`. Then, `c.transactOpts()` will return a pointer to the *copy* of `c.transactOpts` and not to `c.transactOpts`. As a result, any update to the pointer returned by `c.transactOpts()` will only impact the copy of `c.transactOpts` and not `c.transactOpts`. As a result, the update to the `Value` field in the referenced line will only affect a copy of `c.transactOpts` and since we call `c.transactOpts()` to return a pointer to a fresh copy of `c.transactOpts` for each method we don't need to worry about that update affecting other method calls.

**Review by yondonfu — APPROVED** [review]
LGTM after squashing!

File: eth/client_ticketbroker.go (hunk 5 above)
Comment (leszko) [reply eth/client_ticketbroker.go:27]: yes, exactly. Also, the reason for copying `TransactOpts` makes the code thread-safe.
